_posts/2018-07-20-exploring-state-management-in-react-container-components.md

```diff
@@ -0,0 +1,122 @@
+
+      ajax(`https://discountdb.com/?items=${itemsQuery}`)
+        .then(discounts => this.setState({ discounts }));
+    }
+  }
+
+  onNewItemChange = ({target:{value: newItem}}) => this.setState({newItem});
+
+  addItem = () => {
+    this.setState(prevState => ({
+      items: [...prevState.items, prevState.newItem],
+      newItem: ""
+    }));
+  };
+
+  removeItem = item => () => {
+    this.setState(prevState => ({
+      items: prevState.items.filter(
+        (i, idx, prevItems) => idx !== prevItems.indexOf(item)
+      )
+    }));
+  };
+
+  render() {
+    return (
+      <div>
+        <Cart
+          newItem={this.state.newItem}
+          onNewItemChange={this.onNewItemChange}
+          items={this.state.items}
+          discounts={this.state.discounts}
+          addItem={this.addItem}
+          removeItem={this.removeItem}
+        />
+      </div>
+    );
+  }
+}
+```
+[![Edit wo7y9voowk](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/wo7y9voowk)
+&nbsp;
+
+Here you can see that we have a Container component that handles controlling component state with the `addItem`, `removeItem`, and `onNewItemChange` callbacks, and fetching a list of discounts from an external REST api. This enables us to write `Cart` as a Pure Functional component. 
+
+## Benefits
+
+Partitioning your components on their state boundary will help reduce complexity by simply having less code to work with at a time, while still staying "inside React".
+
+I think this pattern really starts to pay dividends when you have a lifecycle method, like `componentDidUpdate`, doing a lot of asynchronous work (like making HTTP requests). Given the asynchronous nature, this sort of code tends to be very difficult to test (with both automated unit testing and manual testing), so breaking this stateful logic into separate components helps keeps you sane and your code focused. 
+
+It's helpful to remind yourself that when unit testing a React component, you are essentially testing the `render` function. Given the inputs (`props`), what is the output? You've probably noticed tests are painful to write if there is a lot of setup, especially if the setup is required for a feature of the component that you aren't even testing.
+
+Keeping your components small and focused will go a long way for keeping yourself happy and productive!
+
+## Drawbacks
+
+While the Container/Presenter pattern is not always one-for-one (Cart <-> CartContainer), you will encounter a lot of similarly named components. This can sometimes cause a communication breakdown amongst the team, as you will trip over your own words attempting to say things like "The CartContainer passes the products to the Cart which then passes them to the Checkout component, or is it the CheckoutContainer component?".
+
+If you can think of better names, I would suggest using them! Your code will still be following the pattern even if they don't have the word Component in the name. 😏
```

Comment:
💡 
> even if they don't have the word Component in the name

I don't think any of your components have `Component` in their name. Did you mean `Container`?